File: music-infrastructure/src/main/java/com/music/sale/persistence/product/entity/ProductItemEntity.java
package com.music.sale.persistence.product.entity;

import com.music.sale.domain.product.enums.*;
import com.music.sale.persistence.common.BaseEntity;
import com.music.sale.persistence.product.support.JsonMapConverter;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.springframework.data.annotation.LastModifiedBy;

import java.util.List;
import java.util.Map;

@Entity
@Table(name = "product_item")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@SuperBuilder
public class ProductItemEntity extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "catalog_id", nullable = false)
    private Long catalogId;

    @Column(name = "seller_id", nullable = false)
    private Long sellerId;

    @Column(name = "store_id", nullable = false)
    private Long storeId;

    //image 추후 업데이트

//    @OneToMany(mappedBy = "productItem", cascade = CascadeType.ALL, orphanRemoval = true)
//    private List<ProductImageEntity> images;

    @Column(name = "name", nullable = false, length = 255)
    private String name;

    @Column(name = "brand", length = 255)
    private String brand;

    // 원 단위 가격 -> BIGINT 매핑
    @Column(name = "price", nullable = false)
    private Long price;

    @Enumerated(EnumType.STRING)
    @Column(name = "condition", nullable = false, columnDefinition = "enum('NEW','USED')")
    private ProductCondition condition;

    @Enumerated(EnumType.STRING)
    @Column(name = "condition_grade", nullable = false,
            columnDefinition = "enum('S','A','B','C','D')")
    private ProductConditionGrade conditionGrade;

    @Column(name = "stock_quantity", nullable = false)
    private Integer stockQuantity;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false,
            columnDefinition = "enum('AVAILABLE','SOLD_OUT','RESERVED','DISCONTINUED')")
    private ProductStatus status;

    @Convert(converter = JsonMapConverter.class)
    @Column(name = "custom_attributes", columnDefinition = "text")
    private Map<String, Object> attributes;

    @Lob
    @Column(name = "description")
    private String description;

    @Column(name = "view_count", nullable = false)
    private Long viewCount;
}

File: music-infrastructure/src/main/kotlin/com/music/sale/persistence/product/entity/ProductItemEntity.kt
// Copyright (C) 2024 Your Name or Company
package com.music.sale.persistence.product.entity

import com.music.sale.domain.product.enum.ProductCondition
import com.music.sale.domain.product.enum.ProductConditionGrade
import com.music.sale.domain.product.enum.ProductStatus
import com.music.sale.persistence.common.BaseEntity
import com.music.sale.persistence.common.JsonConverter
import com.music.sale.persistence.store.entity.StoreEntity
import com.music.sale.persistence.user.entity.UserEntity
import jakarta.persistence.*

/** 제품 아이템 JPA 엔티티 실제 판매되는 상품 인스턴스를 표현 */
@Entity
@Table(name = "product_item")
class ProductItemEntity(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY) val id: Long? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "catalog_id", nullable = false, foreignKey = ForeignKey(name = "none", value = ConstraintMode.NO_CONSTRAINT))
    val catalog: ProductCatalogEntity,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", foreignKey = ForeignKey(name = "none", value = ConstraintMode.NO_CONSTRAINT))
    val seller: UserEntity? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "store_id", foreignKey = ForeignKey(name = "none", value = ConstraintMode.NO_CONSTRAINT))
    val store: StoreEntity? = null,
    @Column(nullable = false) val price: Int,
    @Enumerated(EnumType.STRING) @Column(name = "`condition`", nullable = false) val condition: ProductCondition,
    @Enumerated(EnumType.STRING)
    @Column(name = "condition_grade")
    val conditionGrade: ProductConditionGrade? = null,
    @Column(name = "stock_quantity", nullable = false) val stockQuantity: Int,
    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
    val status: ProductStatus,
    @Column(name = "custom_name") val customName: String? = null,
    @Column(name = "custom_attributes", columnDefinition = "text")
    @Convert(converter = JsonConverter::class)
    val customAttributes: Map<String, Any>? = null,
    @OneToMany(mappedBy = "productItem", cascade = [CascadeType.ALL], orphanRemoval = true)
    val images: MutableList<ProductImageEntity> = mutableListOf(),
) : BaseEntity()
